pkg/guacanalytics/patchPlanning.go

```diff
@@ -0,0 +1,151 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package guacanalytics
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/Khan/genqlient/graphql"
+	model "github.com/guacsec/guac/pkg/assembler/clients/generated"
+	"github.com/guacsec/guac/pkg/misc/depversion"
+)
+
+type DfsNode struct {
+	expanded bool // true once all node neighbors are added to queue
+	Parent   string
+	depth    int
+}
+
+var path []string
+var nodeMap map[string]DfsNode
+var now string
+var nowNode DfsNode
+var queue []string
+
+// TODO: make more robust usuing predicates
+func SearchDependenciesFromStartNode(ctx context.Context, gqlclient graphql.Client, startID string, stopID string, maxDepth int) (map[string]DfsNode, error) {
+	startNode, err := model.Node(ctx, gqlclient, startID)
+
+	if err != nil {
+		return nil, fmt.Errorf("failed getting intial node with given ID:%v", err)
+	}
+
+	_, ok := startNode.Node.(*model.NodeNodePackage)
+
+	if !ok {
+		return nil, fmt.Errorf("Not a package")
+	}
+	queue = make([]string, 0) // the queue of nodes in bfs
+
+	nodeMap = map[string]DfsNode{}
+
+	nodeMap[startID] = DfsNode{}
+	queue = append(queue, startID)
+
+	for len(queue) > 0 {
+		now = queue[0]
+		queue = queue[1:]
+		nowNode = nodeMap[now]
+
+		if stopID == now {
+			break
+		}
+
+		if maxDepth != 0 && nowNode.depth >= maxDepth {
+			break
+		}
+
+		nodes := map[string]model.NeighborsResponse{}
+
+		isDependencyNeighborResponses, err := model.Neighbors(ctx, gqlclient, now, []model.Edge{model.EdgePackageIsDependency})
+		if err != nil {
+			return nil, fmt.Errorf("failed getting package parent:%w", err)
+		}
+
+		nodes["isDep"] = *isDependencyNeighborResponses
+
+		for predicateType, response := range nodes {
+			switch predicateType {
+			case "isDep":
+				err := exploreIsDependency(response, ctx, gqlclient)
+
+				if err != nil {
+					return nil, err
+				}
+			}
+		}
+
+		nowNode.expanded = true
+		nodeMap[now] = nowNode
+	}
+
+	return nodeMap, nil
+
+}
+
+func exploreIsDependency(isDependencyNeighborResponses model.NeighborsResponse, ctx context.Context, gqlclient graphql.Client) error {
+	for _, neighbor := range isDependencyNeighborResponses.Neighbors {
+		if isDependency, ok := neighbor.(*model.NeighborsNeighborsIsDependency); ok {
+			dependentPkgFilter := &model.PkgSpec{
+				Type:      &isDependency.DependentPackage.Type,
+				Namespace: &isDependency.DependentPackage.Namespaces[0].Namespace,
+				Name:      &isDependency.DependentPackage.Namespaces[0].Names[0].Name,
+			}
+
+			depPkgResponse, err := model.Packages(ctx, gqlclient, dependentPkgFilter)
```

Comment:
what about handling this in the outer function? since this is doing exactly that of exploring neighbors?

Reply:
Done.